// gluon-build/dist/commands/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.init = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const __1 = require("..");
const log_1 = require("../log");
const utils_1 = require("../utils");
const init = async (directory, task) => {
    function logInfo(data) {
        if (task) {
            task.output = data;
        }
        else {
            log_1.log.info(data);
        }
    }
    const cwd = process.cwd();
    const dir = (0, path_1.resolve)(cwd, directory.toString());
    if (!(0, fs_1.existsSync)(dir)) {
        log_1.log.error(`Directory "${directory}" not found.\nCheck the directory exists and run |${__1.bin_name} init| again.`);
    }
    let version = (0, fs_1.readFileSync)((0, path_1.resolve)(cwd, directory.toString(), 'browser', 'config', 'version_display.txt')).toString();
    if (!version)
        log_1.log.error(`Directory "${directory}" not found.\nCheck the directory exists and run |${__1.bin_name} init| again.`);
    version = version.trim().replace(/\\n/g, '');
    // TODO: Use bash on windows, this may significantly improve performance. Still needs testing though
    logInfo('Initializing git, this may take some time');
    await (0, utils_1.configDispatch)('git', {
        args: ['init'],
        cwd: dir,
        logger: logInfo,
        shell: 'unix',
    });
    await (0, utils_1.configDispatch)('git', {
        args: ['init'],
        cwd: dir,
        logger: logInfo,
        shell: 'unix',
    });
    await (0, utils_1.configDispatch)('git', {
        args: ['checkout', '--orphan', version],
        cwd: dir,
        logger: logInfo,
        shell: 'unix',
    });
    await (0, utils_1.configDispatch)('git', {
        args: ['add', '-f', '.'],
        cwd: dir,
        logger: logInfo,
        shell: 'unix',
    });
    logInfo('Committing...');
    await (0, utils_1.configDispatch)('git', {
        args: ['commit', '-aqm', `"Firefox ${version}"`],
        cwd: dir,
        logger: logInfo,
        shell: 'unix',
    });
    await (0, utils_1.configDispatch)('git', {
        args: ['checkout', '-b', utils_1.config.name.toLowerCase().replace(/\s/g, '_')],
        cwd: dir,
        logger: logInfo,
        shell: 'unix',
    });
};
exports.init = init;
